Step4: 자동차 경주 (우승자)

programmer-sjk opened the pull request:
안녕하세요 선호님! 
자동차 경주에 우승자 기능을 추가했습니다.

수정된 기능은 아래와 같습니다.
+ **이전 리뷰 반영**
  + lint 반영
  + propery getter 추가
  + 테스트 코드 수정
+ **자동차 이름을 입력값**으로 받아 개수만큼 자동차를 생성
+ **우승자**를 보여주는 코드 추가
+ indent가 2를 넘지 않도록 코드 리팩토링

화요일 수업 전에 4까지는 리뷰를 받고 싶어 부라부랴 작업하게 되었습니다.

programmer-sjk committed:
Step4: 자동차 경주 요구사항 반영

- 자동차 이름을 부여하는 기능 추가
- 우승자를 보여주는 코드 추가
- 인덴트가 2를 넘지 않도록 코드 수정

programmer-sjk committed:
Step4: 자동차 경주 요구사항 반영

- 테스트 코드 수정
- lint 적용
- 프로퍼티 getter 적용

programmer-sjk committed:
Step4: 자동차 경주 요구사항 반영

- README.md 파일 추가

programmer-sjk committed:
Step4: 자동차 경주 요구사항 반영

- 매직넘버 상수화

vsh123 commented on src/main/kotlin/racing/main.kt:
```diff
@@ -2,12 +2,13 @@ package racing
-        println("자동차 대수는 몇 대인가요?")
-        val carNumber = readln().toInt()
+        println("경주할 자동차 이름을 입력하세요")
+        val carNames = readln().split(",")
```
`DELIMITER` 상수화 가능해보여요!

vsh123 commented on src/main/kotlin/racing/Racing.kt:
```diff
@@ -5,27 +5,31 @@ import kotlin.random.Random
+    }
 
-        return distances
+    private fun validateName(name: String) {
```
carName에 대한 validation은 RacingCar가 하면 어떨까요?

vsh123 commented on src/main/kotlin/racing/Racing.kt:
```diff
@@ -5,27 +5,31 @@ import kotlin.random.Random
 
-        return distances
+    private fun validateName(name: String) {
+        if (name.length > MAX_CAR_NAME_LENGTH || name.isBlank()) {
```
[require](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/require.html)를 활용해보세요!

vsh123 commented on src/main/kotlin/racing/RacingCar.kt:
```diff
@@ -1,18 +1,18 @@
 package racing
 
-class RacingCar(private val round: Int) {
+class RacingCar(private val round: Int, val name: String) {
```
`round`는 프로퍼티로 안가지고 있어도 될 것 같아요!

```suggestion
class RacingCar(round: Int, val name: String) {
```

vsh123 replied:
[class layout](https://kotlinlang.org/docs/coding-conventions.html#class-layout)에 맞게 companion object와 프로퍼티들의 순서를 재정렬해보면 어떨까요?

vsh123 commented on src/main/kotlin/racing/RacingCar.kt:
```diff
@@ -1,18 +1,18 @@
-    private val distances = MutableList(round) {0}
-    private var currentRound = 0
+    private val _distances = MutableList(round) { INITIALIZE_VALUE }
+    val distances: MutableList<Int>
+        get() = this._distances
```
불변하게 넘겨주면 어떨까요?

```suggestion
    val distances: List<Int>
        get() = this._distances.toList()
```

vsh123 commented on src/main/kotlin/racing/Racing.kt:
```diff
@@ -5,27 +5,31 @@ import kotlin.random.Random
+        validate(carNames)
+        val racingCars: List<RacingCar> = carNames.map { RacingCar(tryRound, it) }
 
         repeat(tryRound) {
```
> indent(인덴트, 들여쓰기) depth를 2를 넘지 않도록 구현한다. 1까지만 허용한다.

라는 신규 요구사항을 반영해보면 어떨까요?

vsh123 reviewed (changes requested):
안녕하세요!!
step4 구현잘해주셨네요!!

린트확인과 함께 몇가지 코멘트를 남겼으니 확인부탁드릴게요~!

programmer-sjk committed:
Step4: 자동차 경주 요구사항 반영

- 전체 코드에 린트적용

programmer-sjk committed:
Step4: 자동차 경주 요구사항 반영

- 린트 재 적용

programmer-sjk committed:
Step4: 자동차 경주 요구사항 반영

- racing 자동차의 이름 validation을 racingCar에게 책임부여
- split 기준이 되는 쉼표를 상수화
- indent가 2를 넘는 코드 수정
- require를 사용
- MutableList 대신 List를 리턴하도록 수정

programmer-sjk committed:
Step4: 자동차 경주 요구사항 반영

- 프로퍼티 -> 초기화 블록 -> 함수 -> 동반 객체 순으로 정렬

vsh123 commented on src/main/kotlin/racing/RacingCar.kt:
```diff
@@ -22,10 +18,23 @@ class RacingCar(private val round: Int) {
-        distances[currentRound] = MOVE
+        _distances[currentRound] = MOVE
     }
 
```
요 로직 조금 더 살펴보니 처음에 _distances에 init을 할필요가 없어보여요! 한번 고민해보면 좋을 것 같습니다! :)

programmer-sjk replied:
네 아마 아래 코드에서 INITIALIZE_VALUE로 초기화를 할 필요가 없다고 이해했습니다.
초기화 하지 않는 방법으로 수정했는데 맞는지 검토 드립니다. @vsh123 
```kotlin
private val _distances = MutableList(round) { INITIALIZE_VALUE }
```

vsh123 commented on src/main/kotlin/racing/RacingView.kt:
```diff
@@ -2,25 +2,55 @@ package racing
+
+    private fun showRacingWinner(cars: List<RacingCar>) {
+        val cars = cars.map { it.name to it.distances.filter { value -> value == DISPLAY_NUMBER }.size }.sortedByDescending { it.second }
+        val highestScore = cars[HIGHEST_SCORE_INDEX].second
```
`HIGHEST_SCORE_INDEX`가 잘못 설정된것같아요!

programmer-sjk replied:
앗...!! 0이어야 하는데 1로 되어 있군요 수정하였습니다.

vsh123 commented on src/main/kotlin/racing/RacingView.kt:
```diff
@@ -2,25 +2,55 @@ package racing
     }
+
+    private fun showRacingWinner(cars: List<RacingCar>) {
+        val cars = cars.map { it.name to it.distances.filter { value -> value == DISPLAY_NUMBER }.size }.sortedByDescending { it.second }
```
[maxOf](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.comparisons/max-of.html)를 활용해보면 어떨까요?

programmer-sjk replied:
음 선호님 maxOf는 최대값 1개를 리턴 하다보니 
적용하는 과정에서 기존 코드보다 간결하게 작성 한다거나 성능상의 이점? 을 잘 모르겠습니다.

혹시 maxOf 활용을 권유하는 이유나 간결하게 코드를 좀 더 간결하게 만들 수 있을까요..?  @vsh123

vsh123 replied:
성능상에 이점보다는 직접 sorting과 max값의 index를 통해 maxDistance를 안구해도 된다는데에서 오는 장점이 큰 것 같아요!
아래의 예시 코드를 보게되면 sorted과 index없이 maxOf만으로 최대값을 구할 수 있는데요!

이렇게되면 이전 리뷰에서 발견됐던 버그라던지, 코드의 간결함에 한층 더 좋아지지 않을까요~?

<img width="2259" alt="image" src="https://user-images.githubusercontent.com/32254689/169679854-e1e33c05-72ab-414c-a74c-0602c94e248f.png">

programmer-sjk replied:
아하..! 코드를 보고 무슨 말씀인지 이해했습니다. 
감사합니다!

vsh123 reviewed (changes requested):
안녕하세요!
피드백 반영 잘해주셨네요 👍 

한가지 버그가 발견되어 코멘트 남기니 확인부탁드릴게요~!

programmer-sjk committed:
Step4: 자동차 경주 코드리뷰 반영

- 버그 수정 (HIGHEST_SCORE_INDEX 값을 0으로 변경)
- distance 배열을 0 으로 초기화 하지 않도록 변경

vsh123 reviewed (approved):
정국님 안녕하세요!
Step4 구현하느라 고생하셨습니다!

질문에 대해서는 코멘트 남겼으니 확인부탁드릴게요~~~~~~!!

마지막 미션까지 화이팅 하세요!
